pep-0xxx.rst

```diff
@@ -0,0 +1,1470 @@
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+        @abstractmethod
+        def unwrap(self) -> socket.socket:
+            """
+            Cleanly terminate the TLS connection on this wrapped socket. Once
+            called, this ``TLSWrappedSocket`` can no longer be used to transmit
+            data. Returns the socket that was wrapped with TLS.
+            """
+
+
+Buffer
+~~~~~~
+
+The buffer-wrapper ABC will be defined by the ``TLSWrappedBuffer`` ABC, which
+has the following definition::
+
+    class TLSWrappedBuffer(metaclass=ABCMeta):
+        @abstractmethod
+        def read(self, amt: int) -> bytes:
+            """
+            Read up to ``amt`` bytes of data from the input buffer and return
+            the result as a ``bytes`` instance.
+
+            Once EOF is reached, all further calls to this method return the
+            empty byte string ``b''``.
+
+            May read "short": that is, fewer bytes may be returned than were
+            requested.
+
+            Raise ``WantReadError`` or ``WantWriteError`` if there is
+            insufficient data in either the input or output buffer and the
+            operation would have caused data to be written or read.
+
+            May raise ``RaggedEOF`` if the connection has been closed without a
+            graceful TLS shutdown. Whether this is an exception that should be
+            ignored or not is up to the specific application.
+
+            As at any time a re-negotiation is possible, a call to ``read()``
+            can also cause write operations.
+            """
+
+        @abstractmethod
+        def readinto(self, buffer: Any, amt: int) -> int:
+            """
+            Read up to ``amt`` bytes of data from the input buffer into
+            ``buffer``, which must be an object that implements the buffer
+            protocol. Returns the number of bytes read.
+
+            Once EOF is reached, all further calls to this method return the
+            empty byte string ``b''``.
+
+            Raises ``WantReadError`` or ``WantWriteError`` if there is
+            insufficient data in either the input or output buffer and the
+            operation would have caused data to be written or read.
+
+            May read "short": that is, fewer bytes may be read than were
+            requested.
+
+            May raise ``RaggedEOF`` if the connection has been closed without a
+            graceful TLS shutdown. Whether this is an exception that should be
+            ignored or not is up to the specific application.
+
+            As at any time a re-negotiation is possible, a call to
+            ``readinto()`` can also cause write operations.
+            """
```

Comment:
Perversely, `readinto` is not always trivial to implement in a useful way. In particular, CFFI frequently struggles with doing `readinto` because it is often reluctant to allow you to pass a buffer object into C code thanks to PyPy's GC. With many CFFI tools, the reverse is what's done: `readinto` is implemented in terms of `read`.

And given that it is *really easy* to write both by just writing one and then implementing the other in terms of it, there is basically zero cost in specifying that the implementation must provide both, rather than requiring every implementer in the world who wants `read` instead of `readinto` to write their `read` function. =)